R/helpers.R

```diff
@@ -361,21 +368,65 @@ LoadH5AD <- function(path) {
   }
   if (Exists(name = 'raw/X')) {
     md <- 'raw/var'
-    counts <- as.matrix(x = adata[['raw/X']])
+    x <- adata[['raw/X']]
   } else if (Exists(name = 'X')) {
     md <- 'var'
-    counts <- as.matrix(x = adata[['X']])
+    x <- adata[['X']]
   } else {
-    stop("Cannot find counts matrix")
+    stop("Cannot find counts matrix", call. = FALSE)
   }
-  if (!IsMetaData(md = md)) {
-    stop("Cannot find feature-level metadata", call. = FALSE)
-  } else if (!IsMetaData(md = 'obs')) {
-    stop("Cannot find cell-level metadata", call. = FALSE)
+  # check different possible attributes to try and get matrix shape
+  if (isTRUE(x$attr_exists(attr_name = 'h5sparse_shape'))) {
+    mtx.shape <- h5attr(x, 'h5sparse_shape')
+  } else if (isTRUE(x$attr_exists(attr_name = 'shape'))) {
+    mtx.shape <- h5attr(x, 'shape')
+  } else {
+    warning("Could not determine matrix shape")
+  }
+  # check different attributes to try and deduce matrix type
+  if (isTRUE(x$attr_exists(attr_name = 'h5sparse_format'))) {
+    mtx.type <- h5attr(x, 'h5sparse_format')
+  } else if (isTRUE(x$attr_exists(attr_name = 'encoding-type'))) {
+    mtx.type <- substr(h5attr(x, 'encoding-type'), 0, 3)
+  } else {
+    warning("Could not determine matrix format")
+  }
+  if (mtx.type != 'csr') {
```

Comment:
How does this `if (mtx.type != 'csr')` work if the previous checks fail and `mtx.type` is unassigned?

Reply:
in that case it'll assume the matrix is CSR. It's imperfect but without knowing the name of the key storing the matrix type not sure there's a great workaround